src/v/datalake/translation/scheduling.cc

```diff
@@ -0,0 +1,443 @@
+          std::current_exception(),
+          state);
+    }
+}
+
+scheduler::scheduler(
+  size_t total_memory,
+  size_t memory_block_size,
+  std::unique_ptr<scheduling_policy> policy)
+  : _scheduling_policy(std::move(policy))
+  , _mem_tracker(reservations_tracker::make_default(
+      total_memory, memory_block_size, *this)) {
+    ssx::repeat_until_gate_closed_or_aborted(_state.gate, _state.as, [this] {
+        return main().handle_exception([](const std::exception_ptr& e) {
+            if (!ssx::is_shutdown_exception(e)) {
+                vlog(
+                  datalake_log.warn,
+                  "Encountered exception in main loop: {}",
+                  e);
+            }
+        });
+    });
+}
+
+void scheduler::notify_ready(const translator_id& id) {
+    auto holder = _state.gate.hold();
+    vlog(datalake_log.trace, "ready notification from translator: {}", id);
+    auto it = _state.translators.find(id);
+    if (it == _state.translators.end()) {
+        return;
+    }
+    auto& translator = it->second;
+    if (
+      translator._waiting_hook.is_linked()
+      || translator._running_hook.is_linked()) {
+        vlog(
+          datalake_log.warn,
+          "Invalid ready notification from translator, ignoring",
+          translator);
+        return;
+    }
+    vlog(datalake_log.trace, "Marking the translator ready: {}", translator);
+    _state.waiting.push_back(translator);
+    _state_changed_cvar.signal();
+}
+
+void scheduler::notify_done(const translator_id& id) {
+    auto holder = _state.gate.hold();
+    vlog(datalake_log.trace, "done notification from translator: {}", id);
+    auto it = _state.translators.find(id);
+    if (it == _state.translators.end()) {
+        return;
+    }
+    auto& translator = it->second;
+    if (
+      translator._waiting_hook.is_linked()
+      || !translator._running_hook.is_linked()) {
+        vlog(
+          datalake_log.warn,
+          "Invalid done notification from waiting translator, ignoring",
+          translator);
+        return;
+    }
+    vlog(datalake_log.trace, "Marking the translator done: {}", translator);
+    translator._running_hook.unlink();
+    translator._stop_in_progress = false;
+    translator._current_translation_start_time.reset();
+    _state_changed_cvar.signal();
+}
+
+bool scheduler::requires_scheduling_actions() const {
+    return !_state.waiting.empty() || _mem_tracker->memory_exhausted();
+}
+
+void scheduler::notify_memory_exhausted() {
+    vlog(datalake_log.debug, "memory exhausted notification");
+    _state_changed_cvar.signal();
+}
+
+ss::future<> scheduler::stop() {
+    vlog(datalake_log.debug, "Stopping scheduler");
+    _state.as.request_abort();
+    _state_changed_cvar.broken();
+    co_await _state.gate.close();
+    co_await ss::max_concurrent_for_each(
+      _state.translators, 32, [](auto& it) mutable {
+          return it.second._translator->stop();
+      });
```

Comment:
Does this actually need to be run for translators that aren't in the running list?

Reply:
Maybe the translator start()/stop() methods could be called `init()` and `close()` to avoid confusion? Not sure if that's all that much clearer, but at least it avoids confusing the verbs!